### PR TITLE
Add authentication on keystone v3

### PR DESCRIPTION
**Pull request by wjun**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37669)

**Comment by wjun** [comment]
I signed it!

________________________________

File: plugin/pkg/auth/authenticator/password/keystone/keystone.go [hunk 1]
```diff
@@ -39,11 +39,31 @@ type KeystoneAuthenticator struct {
...
 func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticatePassword(username string, password string) (user.Info, bool, error) {
-	opts := gophercloud.AuthOptions{
+	urlSplits := strings.Split(keystoneAuthenticator.authURL, "/")
+	var opts gophercloud.AuthOptions
+	opts = gophercloud.AuthOptions{
 		IdentityEndpoint: keystoneAuthenticator.authURL,
 		Username:         username,
 		Password:         password,
 	}
+	if urlSplits[len(urlSplits)-1] == "v3" {
```
Comment (feiskyer) [inline plugin/pkg/auth/authenticator/password/keystone/keystone.go]: Would it be better to use `strings.HasSuffix()` instead?

**Comment by feiskyer** [comment]
@k8s-bot ok to test

**Comment by feiskyer** [comment]
cc @kubernetes/sig-openstack

**Comment by dims** [comment]
@k8s-bot ok to test

**Comment by xsgordon** [comment]
@idvoretskyi can I get area/platform/openstack or equivalent label please? Thanks!

**Comment by suonto** [comment]
@wjun this is not a generic solution, you're assuming that every user's domain is the end of their email address. This is definitely not always the case. Thus this is a rather non-generic hack and the correct solution for this problem would be to make kubectl read env and pass auth information (icluding domain) to the server which should resolve it in runtime. Furthermore, I think while kubectl is at it, it should just request a token directly from keystone and we could remove keystone auth entirely from the server side. The issue has been discussed in detail here: https://github.com/kubernetes/kubernetes/pull/25391

**Comment by wjun** [comment]
@suonto username@domain is another writing way of domain/username. It is not an email address. Thanks for pointing to that request. It is quite useful to understand the context of this area. I saw some people also would like to keep username/password based authentication during the discussions. This patch may work as a solution to support both keystone v2/v3 username/password authentication.

**Comment by suonto** [comment]
@wjun Thanks for the clarification, I think I understand now what your intention was in this PR. So you suggest that if my username in keystone is `suonto` and I belong to `foobar` domain, I'd configure `suonto@foobar` as my username in kubectl config? That's a good idea, but I'd suggest some other separator than `@` because in enterprises keystone probably retrieves usernames from LDAP or AD and it is almost a standard that enterprises have usernames of format `user.name@enterprise.com`. Therefore, with this solution they'd have to attempt `user.name@enterprise.com@myosdomain`, which of course is not only very confusing but also it'd require a more complex parsing code. However, this solution might indeed be generic if the format was `domain/username` instead of username@domain.

**Comment by wjun** [comment]
@suonto good point, domain or username may include special characters in either way, and I will improve to include this case. For the LDAP/AD cases, I think most users' identification is more like user.name@enterprise.domain, or user.name@myosdomain, isn't it?

**Comment by suonto** [comment]
@wjun As an example of AD or LDAP, if we imagine that my company is eficode and we'd have an openstack, then my company gmail would be markus.suonto@eficode.com and we'd have a user domain `eficode` in OpenStack where our own employees are. Now if we'd have some collaborators from other companies in our openstack, they'd have emails like  user.name@othercompany.com and we'd create them another domain `othercompany` in openstack. Point being, I think the most likely format for federated auth usernames is `firstname.lastname@company.suffix` where `company` is probably used for openstack domain. Anyway, we can't really know how to parse OpenStack user domains from email addresses anyway, so that's why I thought `/` would be a better separator. The case where username contains `/` is much more unlikely to happen than with `@`.

**Comment by zetaab** [comment]
I agree also that / is better than @ in separation. @wjun could you modify that one row userAndDomain := strings.Split(username, "@") -> userAndDomain := strings.Split(username, "/"). Then we could use usernames like "<openstackusername>/<domain>", of course if people have / in their usernames, that might be a problem.

**Comment by grahamhayes** [comment]
Can we not just add a new option, called `Domain_Name` ?

Keystone did not have restrictions on what characters could be used - trying to parse it out would be quite difficult to get correct.

**Comment by suonto** [comment]
@grahamhayes no, domain name must be parsed at runtime. That is, because users can belong to many different domains so if we set one static domain at apiserver start, the users in other domains will get auth failures later when they try to authenticate.

**Comment by wjun** [comment]
@suonto @zetaab The latest commit has changed username to domain\name. @grahamhayes @ttrip This commit will help users to still use username/password approach to authenticate with keystone v3, and not conflict with other keystone authentication proposals. In my opinion, supporting other keystone authentication approaches cannot remove the support of username/password as it is so common one.

**Comment by zetaab** [comment]
@wjun I agree with you, we have kind of similar custom solution in our production environment currently. It is really good feature that some users can still use hardcoded username+password combination and at the same time some users can use keystone authentication.

@grahamhayes suggested that we could add domain_name in some place? Do you mean that we could add that to apiserver arguments? Like @suonto said its not option in this case, because people could have multiple openstack domains in their users.

**Comment by suonto** [comment]
@zetaab I think wjun was speaking of *keystone password auth* not password auth in general.
@wjun I'm not sure if I agree that *keystone password auth* must be preserved in the **server side**. If kubectl client reads OS env variables it could also acquire a token from keystone, then the server side does not need to support username & password, but end users can still use them.

**Comment by wjun** [comment]
@suonto  If we integrate AD into keystone, then kubectl users do not even need to know if a k8s cluster is built on openstack or other IaaS, as they can always use their AD account(domain\name) to login directly. That is one benefit username/password brings to them.

**Comment by suonto** [comment]
@wjun you're completely right. My point was that it is not mandatory to process username and password combinations at the **server side** if we make the **client side** process them earlier. If we make **kubectl** process them and retrieve a token from keystone, ldap, AD, oauth, oidc or whatever other service then the server side only needs to support token auth.

**Comment by liggitt** [comment]
cc @kubernetes/sig-auth-misc

File: plugin/pkg/auth/authenticator/password/keystone/keystone.go [hunk 2]
```diff
@@ -39,11 +39,30 @@ type KeystoneAuthenticator struct {
...
-	opts := gophercloud.AuthOptions{
+	var opts gophercloud.AuthOptions
+	opts = gophercloud.AuthOptions{
 		IdentityEndpoint: keystoneAuthenticator.authURL,
 		Username:         username,
 		Password:         password,
 	}
+	if strings.HasSuffix(keystoneAuthenticator.authURL, "/v3") {
+		// username has a format of domain\name.
+		firstIndex := strings.Index(username, "\\")
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/password/keystone/keystone.go]: can usernames or domains contain a `\` character? if so, we can't assume this is a safe split character.

Comment (suonto) [inline plugin/pkg/auth/authenticator/password/keystone/keystone.go]: You're right. But still, in the current implementation we basically support no v3 auth whatsoever. I think it's a very good improvement to support v3 for all the usernames except those with a backslash.

Comment (zetaab) [inline plugin/pkg/auth/authenticator/password/keystone/keystone.go]: Well actually domain cannot contain \ char now, username can contain?

```
+			dname = username[:firstIndex]
+			uname = username[firstIndex+1:]
```

dname = beginning of string and ends to first \ so if domain name contains \ you have problem.

Comment (liggitt) [inline plugin/pkg/auth/authenticator/password/keystone/keystone.go]: > I think it's a very good improvement to support v3 for all the usernames except those with a backslash.

I'm really hesitant to introduce custom domain encoding schemes to server-side password auth. Letting the client obtain a token directly and present it to the API would be preferable to expanding the username/password support.

Comment (suonto) [inline plugin/pkg/auth/authenticator/password/keystone/keystone.go]: @liggitt I completely agree that the client should do it. But as long as it doesn't (which might be a while), I support this change as a temporary solution. After all, the evaluation of pros and cons is pretty good:
+  no impact on existing functionality as v3 is completely unsupported now
+ support for v3
- does not work for domain names with a backslash
- architecturally not perfect

**Comment by kfox1111** [comment]
Keystone is an authorization system more then an authentication system. Passing username/password to Keystone and then using that strictly for authentication doesn't provide much benefit to just logging in against AD. We bind our Keystone to our AD so all uses have a Keystone account. But not all users have permissions to use the OpenStack. This is via user <-role-> project associations.

The benefit comes when you can SSO with Keystone using Keystone Tokens and then use all the associations of role/project to drive Kubernetes permissions via RBAC.

Keystone supports many different authentication schemes like shibolith, kerberos keytabs, k2k federation and trusts, etc that just won't map well to a username/password world.

I took a stab at implementing this a while back but it stalled out here:
https://github.com/kubernetes/kubernetes/issues/25391
and:
https://github.com/kubernetes/kubernetes/pull/25624

I still think these two ps's are the right approach and in discussion with the Keystone developers they agreed with the approach over using usernames/passwords.

File: plugin/pkg/auth/authenticator/password/keystone/keystone.go (hunk 2 above)
Comment (wjun) [inline plugin/pkg/auth/authenticator/password/keystone/keystone.go]: - The format of "domain\name" is not a new custom domain encoding scheme but widely used by AD based systems where a keystone domain is a AD domain actually. The AD domains usually follow NetBIOS domain name format(https://support.microsoft.com/en-us/kb/909264) which does not allow backslash.
- username/password is useful when  kubectl users do not understand oepnstack and just want to consume k8s(which is hosted on openstack).

**Comment by wjun** [comment]
In my opinion, some k8s end users do not need to understand openstack or whatever IaaS if they just want to consume k8s, even the k8s cluster is deployed on openstack. In this case, only supporting openstack token based auth is not enough, end users still need a well known auth approach such as username/password or openid, etc.

**Comment by zetaab** [comment]
@kfox1111 do you mean that instead of username and password we should use token in our kubeconfig file? Or do you mean that we could still use username and password there but apiserver is making that token automatically for us?

If you are saying that we should use token also in kubeconfig file, its not going to work. Our openstack token lifetime is one hour, it means that everybody should regenerate token all the time. If you are newbie with openstack and k8s do you really want use your time to learn how to generate openstack access token, or do you want just use it? 

If you still allow using username+password in kubeconfig but apiserver will handle tokens, then it is totally different story.

The benefit what we are trying to achieve here is that keystone is making authentication and then we have RBAC which is checking what access this user have. If we want add new user to our k8s cluster, we just edit RBAC file and add new user there. If that user have access to keystone, he can start using k8s by inserting his AD account to kubeconfig file. So also my opinion is that only supporting openstack tokens is not enough.

**Comment by kfox1111** [comment]
All other openstack cli commands use an environment file for dealing with all the settings that are possible to configure appropriate access to keystone. This includes settings such as user name, user domain, project, project domain, keystone auth url, and some kind of credential such as a password or indication that krb tickets should be used, trust info, etc.

The openstack unified cli already has a way to handle all of that info and get a fresh token "openstack token issue". I believe all the kubecli would have to do is have a setting that says the user authentication type is openstack, and then it would shell out to "openstack token issue -f value -c id" and use that with kube-apiserver in token auth mode. It will then function just like any other openstack cli/service. (client gets a fresh keystone token, client talks to service giving it the token, service validates token with keystone and gets back authz info)

No other openstack service takes in a password from the user and gets a token on behalf of the user. Its a security risk to give the password to any service other then the one needing to do the authentication, which is keystone.

**Comment by kfox1111** [comment]
I guess, let me explain the normal user openstack cli workflow a bit further.

User is given access to OpenStack. Usually this comes with a link to the dashboard and some basic instructions on how to use it. This includes a bit of instruction such as:
"If you want to use the OpenStack CLI utilities for accessing your cloud account, login to the web ui, and you can download your cli settings file under Compute->Access & Security->API Access by clicking on "Download OpenStack RC File v3"

All other openstack cli tools accept that file and use it for authentication. It has all the settings prepopulated to talk to the cloud for the user.

I don't think kubecli should function differently then the other cli tools in this workflow. User downloads settings file, sources it, and then can use whichever cli commands that need to talk to the cloud.

**Comment by kfox1111** [comment]
One other bit of information that might be helpful to the conversation. In OpenStack, users arn't directly given permissions to things. They are assigned Roles on Projects. Its the Projects, not Users, that own things. RBAC applies policies based on the Roles a User has on a Project to allow/disallow them operating on a given Project's objects. This allows users to come and go on a Project, as well as Users Roles on a Project to change over time, while ownership of the object is still maintained by the Project. Which can outlast individual users having Roles on a Project.

So authentication against Keystone without project scoping information skips what Keystone really is all about. A single place to manage cloud permissions for all the services.

**Comment by suonto** [comment]
@kfox1111 can we keep a bit of scope on this thread? I've been following the ongoing conversation about keystone auth for a long time, and I do appreciate your work. Also, I'm pretty sure that everyone here agrees that **ideally** the client should process credentials and retrieve a token to present to the server side. However, this PR has been a temporary fix from the beginning; I mean the entire functionality is (hopefully) going to be removed from the apiserver once an ideal client side implementation is ready. The point of this PR though, is to provide v3 **authn** support in the current architecture. Also, I feel like authz really deserves to be a separate thread, as it is more comlicated. For instance kubernetes RBAC also exists, and in my opinion mapping it automatically to keystone RBAC is rather fragile and I'd rather deal with authz in kubernetes and use keystone only for unscoped authn. Anyway, I think the main question here is whether to add or reject this temporary v3 **authn** support or not. So what do you think of that?

**Comment by kfox1111** [comment]
I think the openstack token based authentication (not authorization) plugin patchset meets the requirements of this use case without having the insecure workarounds this ps has. It also eliminates all the discussion around, how do you munge several fields worth of authentication info into just 2 fields (username/password). If someone can pick up that ps real quick and rebase it on trunk, it should be pretty close to done. It was working well when I stopped having time to work on it. All that's missing is a call to openstack token issue in kubectl. manually calling openstack token issue and passing it to kubectl was working fine, as was Kubernetes access from the OpenStack Dashboard. That seems to be a much better path forward to me, as no code will need to be removed to continue to add proper Keystone authorization to Kubernetes in the future, while supporting many more of Keystone's authentication plugins right away.

**Comment by ttripp** [comment]
It seems like this patch doesn't make things worse for password authentication through keystone than they already are, so the scope of its purpose is pretty straightforward.  I think 

That said, @suonto regarding your point on using kubernetes RBAC for authorization once authenticated makes some sense to me (since the permissions in kubernetes don't match with permissions you might manage on openstack projects), but what I don't get is that if you are simply using keystone as a passthrough to AD like @wjun is doing then you won't even know what groups the user belongs to in LDAP / AD. This information doesn't come through on the keystone token returned. Keystone will return roles on projects which possibly could be used to mapped to groups for this user. 

For example, if you want to grant a group of users in AD access to a kubernetes namespace, I don't think that would be possible using this authenticator with the RBAC authorizer (correct)?

I haven't created a diagram like this for AD, but I have created a diagram like this for federated identity which might explain this a bit better:

http://imgur.com/a/QpxwJ

**Comment by wjun** [comment]
@kfox1111 as I and a few guys mentioned in the threads of this patch: token authentication will not remove username/password auth as our k8s users do not want to understand how openstack works but just want to use k8s even it is hosted on openstack. This is good for openstack itself actually as well because users may drop this kind of integrations and pursue other k8s clusters which have a simple widely understood and used auth approach such as username/password or openid. Also, domain\name as username is widely used in AD systems or most companies.

**Comment by wjun** [comment]
@ttripp I do not know how k8s RBAC is implemented towards keystone group/role integration currently but if we pass keystone domain and username information to RBAC, plus openstack configuration information in cloud_config file in k8s api server, RBAC module can invoke openstack keystone apis such as /v3/groups/{group_id}/users/{user_id} to verify a user's group information during applying RBAC rules.

**Comment by suonto** [comment]
@ttripp I remain with my original opinion that **authz** is a separate concern, and mappings are part of authz.

@wjun removing the username/password auth is rather transparent to end users. Also, it is definitely a better architecture to process the credentials into a token at the client side so I think your argument stands on thin ice.

I think I stand with @kfox1111 on this, we should just focus our efforts towards the proper implementation. However, as this implementation is a harmless addition, it doesn't really matter in my opinion if it's merged or closed.

**Comment by wjun** [comment]
It looks we have enough discussions on the related backgrounds of this patch and openstack/keystone and we all want openstack based k8s better. I would suggest to focus on this patch itself again as it first appears.

**Comment by kfox1111** [comment]
To speak to the non authz case, I still think the other patch set does it both properly, and covers the use case better. The user gets their openstack rc file as normal, and then can start using kubectl as if it were any other openstack command. No special openstack configuration just for kubectl is required. No special rules to munge keystone credential information into k8s is required and seems counter intuitive as a user. It supports more Keystone Authentication methods. It is a base to build on for further integrations. It does things the way Keystone expects things to be done.

To the authz group question,
Yeah, keystone does have a concept of groups but it is opaque to any other service. Users/Groups of users get assigned roles on projects. A user authenticates with username/password/domain info and project and gets a scoped token back, and a list of active roles on the project. these roles on the project are whats expected to be used for authz, not any group membership. And the user is acting on behalf of the project authenticated against, not him/her self. so the unit of ownership is the Project, not the user. If that user at any moment tries to access a different project then scoped to, they are denied access, even though they may have access to that project if they get a scoped token for that project. Only a single project can be acted on at a time per token. This prevents a user with multiple projects from accidentally making changes to the wrong project.

Keystone is not about authentication at all. It just has a minor authentication component in it, and they are actually trying to get rid of it. If you really want username/password authentication with group membership back, that is what AD/LDAP were built for. Keystone was built for something very different. Multitenant management/authorization.

**Comment by kfox1111** [comment]
@wjun I think the problem is you think because keystone is using the same terminology as AD, username/password/domain, you think it functions like AD. It does not.

This patch set will also exacerbate that problem, as it perpetuates the misunderstanding of how openstack works.

It also poses an increased security risk as its passing around credentials to places that don't need them.

The way all other openstack cli tools work is not by username/domain/password, but by environment file based config. Your arguing for doing something not normal for openstack commands and trying to use keystone in a way it wasn't built for. It won't be easier to use for openstack users as it functions different then the other tools and OpenStack administrators go through efforts to setup cli environments for their users so in those cases, the token based authentication will "just work". requiring users to tweak their kube config with values that they have to lookup through some other undefined mechanism is less intuitive, not easier to use.

My point is, users don't have to understand how openstack auth works if implemented properly. With this PS, they do.

**Comment by zetaab** [comment]
@kfox1111 at least i do not even want that those users who have access to kubernetes (using keystone tokens) have access also to openstack tenant. It looks like that it is what you are trying to achieve here? Lets imagine situation that I am big k8s cluster admin which is running in top of openstack. Do I want that every single user have access to openstack tenant as well where they can play with k8s host machines? No. I want that openstack and k8s is separated things. With this PR and RBAC we can achieve that.

Then second thing: it looks like you are just fighting against this PR and saying ideas which is better. But the main problem here is that at least I do not see any other PR which contains those things what you have mentioned? You should understand that 1) currently there is NO support for keystone v3 at all 2) doing that your PR takes time (i think months-years because i have been following this keystone issue quite long time). 

If we merge this PR is it going to crash your world? Not really, you can still do that your superhyper PR and change our world. Please do it and before that let us use even something from keystone.

Thanks

**Comment by liggitt** [comment]
I agree that token auth is preferable, for many reasons (security, best matched to openstack ecosystem, supports many different auth methods), and would like to see progress made in https://github.com/kubernetes/kubernetes/pull/25391

For this PR, I would not object to adding support for keystone v3 for a single domain, but do not want to invent a username+domain encoding scheme specific to kubernetes. Among other reasons, it makes the username for a user in the default domain ambiguous (could be `bob\default` or just `bob`).

**Comment by zetaab** [comment]
It could be `default\bob` or just `bob` for default domain. But how that differs from AD/LDAP things what people are using nowadays? Usually you can authenticate against those systems 1) using only your username or 2) using `domain\username`.

**Comment by kfox1111** [comment]
@zetaab the ps was referenced above, but here it is again for reference:
https://github.com/kubernetes/kubernetes/pull/25391
Part of the problem getting that patch set through was its size due to trying to continue to support the existing v2 username/password based auth bits. Accepting this PS would probably mean further impediments to merging the other patch set.

I'm suggesting, that I believe the proper way is that if you want to integrate keystone, the users would be able to login with their existing keystone logins and the rbac system would have enough knowledge of keystone to use the existing rbac info for those users from keystone to not require lots of special, per user k8s rbac rules, but rules that operate per project/role. You set up the user once in keystone like you normally do, and they get restricted access to their subset of the k8s system. I'm not arguing the user gets access to the tenant k8s is running in.

Its not just k8s rbac or keystone rbac, but both working together that makes the system easy to maintain. That is how all other openstack services work. keystone roles (rbac) ties in with per service policies (eqiv to k8s rbac) to provide a complete, scalable, multi tenant aware authz system.

**Comment by suonto** [comment]
@liggitt agreeing with @zetaab, it could be just username for default domain and domain\username for other domains. A single domain implementation is strictly inferior. Also, I think it would infuriate openstack users when they realise that multi-domain is not supported on purpose. At least for me the patch would be useless without multi-domain support.

@kfox1111 Even if this is merged, it doesn't really conflict with #25391, which I really would like to see completed as well. I'd help you with that, but don't really have the time now.

**Comment by zetaab** [comment]
@kfox1111 and @liggitt when you are talking about openstack token etc, does it mean that every user should have access to tenant x in order to access k8s cluster? If yes, its not going to work. Everybody cannot have access to k8s host machines.

**Comment by kfox1111** [comment]
no. users do not need access to the tenant the service is running out of.

We do this all the time with services like desigate, where you may run the designate service from within the "service" tenant in the same openstack its part of, and then all openstack users on that cloud can use it but they don't have access to the the "service" tenant. That is really what Keystone is all about. Tenant authorization and isolation.

**Comment by liggitt** [comment]
> @liggitt agreeing with @zetaab, it could be just username for default domain and domain\username for other domains. A single domain implementation is strictly inferior. Also, I think it would infuriate openstack users when they realise that multi-domain is not supported on purpose. At least for me the patch would be useless without multi-domain support.

@suonto since user and domain names can contain `\`, `@`, and all sorts of other characters, there's not an unambiguous way to combine them into a single string. That might not matter in your use case, but it's not a good idea to expand in-tree auth integrations beyond what we can ensure correctness on.

@wjun short-term, if scoping this PR down to just add v3 support for a single (optionally configurable) domain could be useful, I think that could be ok. long-term, I'd like to see the token support added server and client-side and drop the experimental password support.

**Comment by wjun** [comment]
@liggitt The format of "domain\name" in AD domains usually follow NetBIOS domain name format(https://support.microsoft.com/en-us/kb/909264) which does not allow backslash. I think it should satisfy most keystone v3 users' naming requirements for now.

**Comment by zhouhaibing089** [comment]
Long thread, I read some of the comments, in case guys does not know the `auth_provider`, please see this as an implementation: https://github.com/kubernetes/kubernetes/pull/39587

Client-side token generation is totally possible, there is a hook point when the client can get credentials from, like environment variables. and in this case, client side can generate a token, or even rotate automatically.

**Pull request closed without merging**